### ISSUE TITLE
[WOTLK] Raise Dead Spell for Death Knight

### ISSUE THREAD

**Issue by Tauriella** [cmangos/issues#709]
Raise Dead does not works 
if you use this spell nothing happens

http://www.wowhead.com/spell=46584/raise-dead

**Comment by xfurry** [comment]
Poke @kvipka

**Comment by Tauriella** [comment]
Yea Poke @kvipka and @xfurry  :+1:

**Comment by xfurry** [comment]
I found this on R2. Not sure if it's right or wrong, as I don't know much about DKs.

```c++
// Raise dead script effect
                case 46584:
                {
                    if (!unitTarget || m_caster->GetTypeId() != TYPEID_PLAYER)
                        return;

                    // If have 52143 spell - summoned pet from dummy effect
                    // Another case summoned guardian from script effect
                    uint32 triggered_spell_id = m_spellInfo->CalculateSimpleValue(SpellEffectIndex(m_caster->HasSpell(52143) ? EFFECT_INDEX_2 : EFFECT_INDEX_1));

                    float x,y,z;

                    m_caster->GetClosePoint(x, y, z, m_caster->GetObjectBoundingRadius(), PET_FOLLOW_DIST);

                    if (unitTarget != (Unit*)m_caster)
                    {
                        m_caster->CastSpell(unitTarget->GetPositionX(),unitTarget->GetPositionY(),unitTarget->GetPositionZ(),triggered_spell_id, true, NULL, NULL, m_caster->GetObjectGuid(), m_spellInfo);
                        //if (unitTarget->GetTypeId() == TYPEID_UNIT) // By user information, corpse not removed after ghoul summon.
                        //    ((Creature*)unitTarget)->RemoveCorpse();
                    }
                    else if (m_caster->HasAura(60200))
                    {
                        m_caster->CastSpell(x,y,z,triggered_spell_id, true, NULL, NULL, m_caster->GetObjectGuid(), m_spellInfo);
                    }
                    else  if (((Player*)m_caster)->HasItemCount(37201,1))
                    {
                        m_caster->CastSpell(m_caster,48289,true);
                        m_caster->CastSpell(x,y,z,triggered_spell_id, true, NULL, NULL, m_caster->GetObjectGuid(), m_spellInfo);
                    }
                    else
                    {
                        SendCastResult(SPELL_FAILED_REAGENTS);
                        finish(true);
                        CancelGlobalCooldown();
                        return;
                    }
                    finish(true);
                    m_caster->RemoveSpellCooldown(m_spellInfo->CalculateSimpleValue(EFFECT_INDEX_2),true);
                    m_caster->RemoveSpellCooldown(m_spellInfo->CalculateSimpleValue(EFFECT_INDEX_1),true);
                    CancelGlobalCooldown();
                    return;
                }

                // Raise ally
                case 61999:
                {
                    if (m_caster->GetTypeId() != TYPEID_PLAYER)
                        return;

                    if (!unitTarget || unitTarget->GetTypeId() != TYPEID_PLAYER || unitTarget->isAlive())
                    {
                        SendCastResult(SPELL_FAILED_TARGET_NOT_DEAD);
                        finish(true);
                        CancelGlobalCooldown();
                        return;
                    }

                    // hack remove death
                    unitTarget->CastSpell(unitTarget, m_spellInfo->CalculateSimpleValue(EFFECT_INDEX_0), true);
                    CancelGlobalCooldown();
                    return;
                }
```

**Comment by Grz3s** [comment]
im not an expert with spells ... but this looks promissing :)..
maybe @kvipka can look at his code? I know u worked on these spells.
I'll look at this code .. anyway ;)

**Comment by Argonn** [comment]
We did some testing with the code from @xfurry, its works wonders. But some minor things still looks weird.
1. Still cannot be used, with a humanoid corpse without corpsedust (a corpse that gave XP ! )
2. The cooldown will start to tick from the moment that you'r ghoul dies. (unclear about this one). But maybe this is because I'm unholy with permanent Ghoul.

And for the code:
1. We changed m_caster->RemoveSpellCooldown to ((Player*)m_caster)->RemoveSpellCooldown, otherwise the code would not compile.

**Comment by Argonn** [comment]
Strangly you can also use 'living' creatures as substitute for dust. Casting the spell next to "Squire Edwards" within DK area, it will show under him, and then summon your ghoul without wasting any dust.

**Comment by Mightylink** [comment]
I'm so surprised this is still bugged, we've been wanting ghouls since wotlk first came out.

**Comment by Vladskywolf** [comment]
I haven't seen this mentioned, so I'll add this to the discussion. 

Ghouls can be raised by manually casting spell 46585 (.cast 46585) instead of 46584 (the trained spell id) so they do work and this can conveniently be put into a macro. However, this is not an ideal way to do it because:

- 'cast' command permission needs to be changed in the database to 0 for all players (normally 3).

- it bypasses checks for corpse dust and/or viable humanoid corpses.

- any class will have the ability to summon a ghoul, which of course is not good.

- it bypasses the 3 minute cooldown so you'd be able to summon ghouls to your heart's content.

All in all, the list of disadvantages to doing it this way is too great, unless of course you have a small trusted playerbase which won't abuse it, or you're the GM. ;) However _something_ in the code is working or I shouldn't be able to summon one at all I would think.

Interestingly, I have a DK which I created in a much older build (12576) which can summon a ghoul using the spell from the spell book (not the macro). However, I can only cast it once, then I'd have to logout/login to use it again, and it displays a small green targeting circle first rather than instantly calling it forth. Any DK created in later versions I've tested and/or used (12938, 12940, 12942, 12943, 13956) doesn't successfully cast this at all with the trained version of the spell.

**Comment by Vladskywolf** [comment]
It's been a while since I tested any new builds (Jul. 2017) but I decided to try the latest build and I can say something has changed, unfortunately for the worse.

As I mentioned previously, I could call forth a ghoul casting a spell id via a macro. This still works, but only once whereas before I could use it throughout my play session. Now it tells me that I can't cast that spell yet whenever I try again, despite waiting well past the designed three minute cooldown.

**Commit Hash**:
9b98854f15751d08aa65d32b96abf92eb8a683f1

**Database**:
WoTLKDB v1.1 'The Warsong' for CMaNGOS-WOTLK 13965
Hash: a9fedfaf62188a86f904b16389df284b82d60afd
